### ISSUE TITLE
New data structure for Sample Exam

### ISSUE THREAD

**Issue by danopolan** [istqborg/istqb_product_base#70]
Recent delivery of [CTAL TA Sample Exam](https://github.com/istqborg/istqb-ctal-ta/blob/alpha-dev-exam/sample-exam/questions.yml) for Alpha review raised some questions on using YAML format for all sample questions. It's working fine for simple questions like we mostly have in [CTFL module](https://github.com/istqborg/istqb-ctfl/blob/main/sample-exam/questions-a.yml), but for all advance modules, this format is becoming hard to use.

Specifically, the issues are:

- Authors do not have Markdown preview while creating questions
- Authors make mistakes in indents while working with YAML file
- DOCX output is just YAML in DOCX format

I was thinking about splitting questions into separate Markdown files, with the structure like below. Then files can be parsed and merged into a single YAML file within the pipeline.

```markdown
# metadata
lo: 1.1.1
k-level: K1
points: 1
correct: c

## question 
Which of the following statements describe a valid test objective?

## answers
1. To prove that there are no unfixed defects in the system under test
2. To prove that there will be no failures after the implementation of the system into production
3. To reduce the risk level of the test object and to build confidence in the quality level
4. To verify that there are no untested combinations of inputs

## justification
1. Is not correct. It is impossible to prove that there are no defects anymore in the system under test. See testing principle 1
2. Is not correct. See testing principle 7
3. Is correct. Testing finds defects and failures which reduces the level of risk and at the same time gives more confidence in the quality level of the test object
4. Is not correct. It is impossible to test all combinations of inputs (see testing principle 2)
```

But this is just a task to discuss the approach.

**Comment by danopolan** [comment]
I am collecting feedback on this one from engineers, stakeholders and authors. Just you know that this is being worked on.

**Comment by danopolan** [comment]
So the preferred solution by various people is to keep questions text on Markdown using specific structure and parsing it as part of CI into YAML.

This approach is key for:
* Easy verification of text, tables and images rendering of MD in the Markdown editors
* DOCX output which is key for language reviews and translations for some member boards
* Less errors in YAML format while written manually
* Ability to split questions into separate files by question, chapter or keep them all in one.

So I see several things we need to do here:
- prepare the file structure in Markdown for each question
- create script to parse MD files into YAML
- test and retrofit this to existing questions

This is still for discussion, but the users have spoken :)

**Comment by danopolan** [comment]
@Witiko can we agree on next steps and format of Markdown for this issue?
I would like to start changing current TA Sample Exam question to a new form in parallel, so we can allow language reviewers to start working on it.

**Comment by Witiko** [comment]
> @Witiko can we agree on next steps and format of Markdown for this issue?

Sure. The format in https://github.com/istqborg/istqb_product_base/issues/70#issue-2345726246 seems appropriate for a machine conversion to YAML.

**Comment by Witiko** [comment]
@danopolan Do we have any actual question definitions in the new format for testing the implementation or should by start by retrofitting existing questions?

In `questions.yml`, there are additional values that apply to all questions:

https://github.com/istqborg/istqb_product_base/blob/9b00830bd9ca9348a5022486a127e5bb1c34d7d9/schema/questions.yml#L1-L4

We can determine automatically the field `num-questions` but how would we determine the other fields? Furthermore, since we don't currently use the fields in the sample exam documents, we can also make them optional in the schema and not generate them for the moment.

**Comment by danopolan** [comment]
> Do we have any actual question definitions in the new format for testing the implementation or should by start by retrofitting existing questions?

Yes, there is [questions.md](https://github.com/istqborg/istqb-ctal-ta/blob/questions_redesign/sample-exam/questions.md) in ctal-ta.

> In `questions.yml`, there are additional values that apply to all questions:

Yes, we can omit these for now, since we do not use them. I would just remove them for now.
